[WebGPU] Abandoned or failed command buffers can workaround some validation

mwyrzykowski opened the pull request:
#### fefb539ed6447b7df288366d26d52813e7a693e9
<pre>
[WebGPU] Abandoned or failed command buffers can workaround some validation
<a href="https://bugs.webkit.org/show_bug.cgi?id=286578">https://bugs.webkit.org/show_bug.cgi?id=286578</a>
<a href="https://rdar.apple.com/143628699">rdar://143628699</a>

Reviewed by Cameron McCormack.

We need to check for status == completed prior to assuming the result
in the auxillary buffer contains reliable information.

For instance, a command buffer which was dropped and had a zero ref
count could trigger the completion handler to run without the work
ever being submitted to the MTLCommandQueue.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::checkForIndirectDrawDeviceLost):
(WebGPU::RenderPassEncoder::executeBundles):

* LayoutTests/fast/webgpu/nocrash/fuzz-286578-expected.txt
* LayoutTests/fast/webgpu/nocrash/fuzz-286578.html
Add regression test.

Canonical link: <a href="https://commits.webkit.org/289519@main">https://commits.webkit.org/289519@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9430b6e02010c0f48cfd2113d8dea5a26f865c7e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41400 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91903 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37784 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6831 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14623 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67277 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25039 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90047 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5232 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78784 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47599 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5008 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33159 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36901 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75500 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34036 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93791 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10339 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76199 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14411 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75284 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18552 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19624 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18045 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7124 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14226 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19519 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13970 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17413 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15751 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam reviewed (approved):
There are a few other uses of `-[MTLCommandBuffer addCompletedHandler:]` in other files. Do they need any checks for their status?

* The one in Queue.mm is already checking for the status
* The one in CommandBuffer::makeInvalidDueToCommit it probably makes sense it doesn't need to check
* The one in Device::resolveTimestampsForBuffer I'm not sure about

mwyrzykowski committed:
[WebGPU] Abandoned or failed command buffers can workaround some validation
https://bugs.webkit.org/show_bug.cgi?id=286578
rdar://143628699

Reviewed by Cameron McCormack.

We need to check for status == completed prior to assuming the result
in the auxillary buffer contains reliable information.

For instance, a command buffer which was dropped and had a zero ref
count could trigger the completion handler to run without the work
ever being submitted to the MTLCommandQueue.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::checkForIndirectDrawDeviceLost):
(WebGPU::RenderPassEncoder::executeBundles):

* LayoutTests/fast/webgpu/nocrash/fuzz-286578-expected.txt
* LayoutTests/fast/webgpu/nocrash/fuzz-286578.html
Add regression test.

Canonical link: https://commits.webkit.org/289519@main
